src/main.ml

```diff
@@ -1,488 +1,323 @@
-      Lwd_utils.pack Ui.pack_x
-        [
-          place_ui_var left_pane ~sw:1;
-          spacer;
-          place_ui_var middle_pane ~sw:2;
-          spacer;
-          place_ui_var right_pane ~sw:6;
-        ];
-      place_ui_var footer;
-    ]
+    [ place_ui_var header; Lwd.get body |> Lwd.join; place_ui_var footer ]
 
-let errorf fmt = Printf.ksprintf failwith fmt
+let render_list_item highlight text =
+  let attr =
+    if highlight then Notty.A.(st bold ++ fg lightblue ++ st reverse)
+    else Notty.A.(st bold ++ fg lightblue)
+  in
+  NW.string ~attr text
+
+let render_state highlight (state : Raw.Reader.JobInfo.State.unnamed_union_t) =
+  let icon, color =
+    let open Notty.A in
+    match state with
+    | NotStarted -> ("[ ]", white)
+    | Passed -> ("[✓]", green)
+    | Failed _ -> ("[X]", red)
+    | Active -> ("[.]", yellow)
+    | Aborted -> ("[A]", lightred)
+    | Undefined _ -> ("[�]", white)
+  in
+  if highlight then NW.string ~attr:Notty.A.(fg color ++ st reverse) icon
+  else NW.string ~attr:Notty.A.(fg color) icon
 
 let import_ci_ref ~vat = function
   | Some url -> Capnp_rpc_unix.Vat.import vat url
   | None -> (
       match Sys.getenv_opt "HOME" with
-      | None -> errorf "$HOME not set! Can't get default cap file location."
+      | None -> failwithf "$HOME not set! Can't get default cap file location."
       | Some home ->
           let path = Filename.concat home ".ocaml-ci.cap" in
           if Sys.file_exists path then Capnp_rpc_unix.Cap_file.load vat path
-          else errorf "Default cap file %S not found!" path )
-
-let with_ref r fn =
-  Lwt.finalize
-    (fun () -> fn r)
-    (fun () ->
-      C.Capability.dec_ref r;
-      Lwt.return_unit)
-
-let show_log table job =
-  let add str = if str <> "" then Lwd_table.append' table str in
-  let rec aux start =
-    Current_rpc.Job.log ~start job >>= function
-    | Error _ as e ->
-        add "ERROR";
-        Lwt.return e
-    | Ok ("", _) ->
-        add "DONE";
-        Lwt.return_ok ()
-    | Ok (data, next) ->
-        add data;
-        aux next
-  in
-  aux 0L
+          else failwithf "Default cap file %S not found!" path )
 
-let control_character_index str i =
-  let len = String.length str in
-  let i = ref i in
-  while
-    let i = !i in
-    i < len && str.[i] >= ' '
-  do
-    incr i
-  done;
-  if !i = len then raise Not_found;
-  !i
+let open_in_editor var log_lines = function
+  | `Activate ->
+      let fname, oc = Filename.open_temp_file "citty" "log" in
+      Lwd_table.iter (output_string oc) log_lines;
+      close_out_noerr oc;
+      ( try
+          let safe_name = Filename.quote fname in
+          let candidates =
+            match Sys.getenv_opt "EDITOR" with
+            | Some x -> [ x ]
+            | None -> (
+                match Sys.getenv_opt "PAGER" with
+                | Some x -> [ x ]
+                | None -> [] )
+          in
+          let candidates = candidates @ [ "xdg-open"; "open" ] in
```

Comment:
`$VISUAL` should be checked before `$EDITOR`, it should be a terminal editor too.

On macos, `open` terminates immediately, sometimes before the file is loaded by the editor.